### ISSUE TITLE
Some feedback

### ISSUE THREAD

**Issue by martijnthe** [MiguelGuthridge/Jestspectation#72]
Hi Miguel,

First off, thanks for writing this library!
I found it through your SO post, after googling something like "jest for python".
I've been playing with the library this afternoon and even though I'm probably not going to use it, I wanted to give you my feedback and encourage you to keep improving your lib! So here it goes...

When I saw [this example](https://miguelguthridge.github.io/Jestspectation/#usage-with-pytest):
```
    def test_lists():
>       assert expect.Equals([1, 2, 3, 4]) == [1, 2, 3, 5, 6]
E       assert [1, 2, 3, 4] == [1, 2, 3, 5, 6]
E         !! [3] 4 == 5
E            Value mismatch
E            Expected 4
E            Received 5
E         ++ [4] 6
```

I thought, "why, this is not better than pytest's builtin diffing" -- when running `pytest -vv`:
```
    def test_lists() -> None:
>       assert [1, 2, 3, 4] == [1, 2, 3, 5, 6]
E       assert [1, 2, 3, 4] == [1, 2, 3, 5, 6]
E         At index 3 diff: 4 != 5
E         Right contains one more item: 6
E         Full diff:
E         - [1, 2, 3, 5, 6]
E         ?           ^^^^
E         + [1, 2, 3, 4]
E         ?           ^
```

Next, I was interested in what error reporting would look like with a mismatch happening in a nested matcher:
```
    def test_foo() -> None:
>       assert [
            "extra but fine",
            {
                "tux": ["bad!"],
            },
        ] == expect.ListContaining(
            [
                expect.DictContainingItems(
                    {
                        "tux": expect.ListContainingOnly(["foo"]),
                    }
                ),
            ]
        )
E       AssertionError: assert ['extra but fine', {'tux': ['bad!']}] == ListContaining([...])
E         Missing properties
E         Expected a ListContaining([...])
E         -- DictContainingItems({...})
```
I was hoping it would call out that the `["bad!"]` list did not contain `"foo"`, but it doesn't.
(I understand why it works like it does, but from a user perspective, this is not the most helpful thing.)
Also, "Missing properties" could be more helpful. What properties?
(Note the "properties" here are the list items.)
The elision (...) is not helpful. I think eliding matchers / parts that matched OK is perhaps (?) fine, but I would surely not elide failing matchers.

I was curious whether pytest's diffing couldn't be leveraged.
So, I tried a quick 'n dirty "contains dict" matcher, that subclasses `dict` itself, so pytest's diffing kicks in per usual:

```
class DictWithItems(dict):
    def __eq__(self, other):
        if not isinstance(other, dict):
            return False
        for key, value in self.items():
            if key not in other:
                return False
            if value != other[key]:
                return False
        return True


def test_foo() -> None:
>       assert [
            "extra but fine",
            [
                {"deep": "bad!", "extra": 123},
            ],
        ] == [
            "extra but fine",
            [
                DictWithItems({"deep": "good"}),
            ],
        ]
E       AssertionError: assert ['extra but fine', [{'deep': 'bad!', 'extra': 123}]] == ['extra but fine', [{'deep': 'good'}]]
E         At index 1 diff: [{'deep': 'bad!', 'extra': 123}] != [{'deep': 'good'}]
E         Full diff:
E         - ['extra but fine', [{'deep': 'good'}]]
E         ?                               ^^^
E         + ['extra but fine', [{'deep': 'bad!', 'extra': 123}]]
E         ?                               ^^ + ++++++++++++++
```
Unfortunately, the differ also highlights the `'extra': 123`, which we're not interested in because it wasn't part of the matcher's input. I browsed through pytest's code, but this is a dead end because the diffing is rather "dumb": it happens on the `repr` text strings of the expectated & actual values, not the original tree structure.

I think to solve this, the diffing would need to be done "semantically". [This lib](https://github.com/trailofbits/graphtage) might be useful there (came up when googling for semantic diffs).

Finally some minor notes:
- I would add one or more examples to each API in the docs.
- I'd add link to docs to README.md. I saw there's a docs link in GH in the side bar, but it feel like the README.md is more prominent / easier to discover.
- Why use `ListContainingOnly` instead of a plain list `[]`?
- Doc for `Equals` is incorrect "Matches values that have the same identity as the given value."

Anyway, I hope you're not discouraged by all of this, but instead find inspiration to improve your lib further!

**Comment by MiguelGuthridge** [comment]
Thanks so much for the feedback! These are all things that could definitely benefit from some improvement. 

* I completely agree that Pytest's diff matching is a little nicer (and presumably less buggy), so improving the way that the diffs are written is absolutely a good idea!
* Particularly for simple cases, Pytest's short explanations are very nice - it could be helpful to determine the level of explanation to give using the "depth" of the data structures.
* Semantic diffing could be a good start, but I think the complexity of some of the matchers may demand the code be written as part of this library

I've opened #73, where I'll work on improving this when I next have time/energy/motivation.

To answer your other questions/notes

* Adding examples to the docs: awesome idea, I'll do this as soon as I get the time, I've opened #74 as a reminder to myself!
* Adding docs link to the readme: done!
* Why use  `ListContainingOnly`: this matches data in any order (kind of like a set) - helpful, since matching on a `set` will fail due to other type checking `[1,2, 3] != {1, 2, 3}`
* Docs for `Equals` were wrong: fixed :)

Thanks again for all the feedback! Sorry I can't fix all of it immediately, but I'll work on it slowly over time. Let me know if there are any other improvements that can be made!
